Localizable craftmenu

Ertanic opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
The name, description and icon of the target crafting items are now taken directly from the prototypes. The ability to specify a name is left through localization files. See the secret door for an example.

Depends on ~~https://github.com/space-wizards/RobustToolbox/pull/5466~~.

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->

Fixes #22814

## Technical details
<!-- Summary of code changes for easier review. -->
Because prototypes are read only on the server, I moved some of the code from the Server segment to Shared and Client. When ConstructionPrototype comes in, it creates a ConstructionPrototype.ID-to-EntityPrototype.ID map, which the client then works with.

To take sprites directly from the prototypes, `ItemList` has been replaced by `ListContainer`, which uses `EntityPrototypeView`.

Fluent keys have been used to localize GraphStep. Therefore, the names of the `tag` and `material` steps, which are used directly in the build menu, have been moved to ftl files.

The `MaterialConstructionGraphStep` (set in StackPrototype) passes the necessary amount of material to be translated using selectors. Especially relevant for the Slavic language group.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
- tweak: The name, description and icon of the target crafting items are now taken directly from the prototypes.

slarticodefast commented:
Wouldn't that cause problems with things like secret walls that have a different name?

Ertanic commented:
> Wouldn't that cause problems with things like secret walls that have a different name?

Very likely. But exactly the secret door in the prototype itself is referred to as a solid wall. 

![image](https://github.com/user-attachments/assets/5d7b8d57-626a-40e6-a45f-0c47762303fc)

slarticodefast commented:
Yeah, that's a problem, you won't know which one is which if you don't know the recipe. Maybe add an optional translation string to the crafting menu entry that overrides the prototype name if necessary.

chavonadelal commented:
> Yeah, that's a problem, you won't know which one is which if you don't know the recipe. Maybe add an optional translation string to the crafting menu entry that overrides the prototype name if necessary.

No, I don't think that's an option! So localization could be ruined again... Although, only if it's LocID, then maybe...

Ertanic commented:
> Yeah, that's a problem, you won't know which one is which if you don't know the recipe. Maybe add an optional translation string to the crafting menu entry that overrides the prototype name if necessary.

I left the ability to overload prototype craft names manually via fluent keys. Among other things, I added an optional ability to output suffixes. 

At the request of @chavonadelal I made it possible to pull sprites from target prototypes. I didn't bother and just replaced ItemList and TextureRect by ListContainer and EntityPrototypeView. Then I will remove unused icon fields in construction prototypes.

![image](https://github.com/user-attachments/assets/42afcfc8-2f1a-4e50-9d8c-b2cada96b7b9)

![image](https://github.com/user-attachments/assets/09f63881-758f-41c5-b3e7-8ee9d5e94a85)

mirrorcult commented:
extremely blessed PR

lzk228 commented:
> I left the ability to overload prototype craft names manually via fluent keys. Among other things, I added an optional ability to output suffixes.

pjb once said
![image](https://github.com/user-attachments/assets/57b37e9a-5204-4f1b-bade-7014b8763b2d)

Ertanic commented:
Sneaky rider decided to brush up on unused imports.....

metalgearsloth commented:
so:
- I got randomly pinged over this PR merge-begging for it when it was only a few days old.
- I look at the engine PR and it's bandaiding a bug and 2 maintainers ask for it to be investigated (whether it's tests or control-related).
- Instead you tell us it's someone else's problem and refuse to fix it and we should prioritise your PR instead as someone else can fix the bug.

If you keep insisting our engine is so bad is the better solution not to investigate the root causes of issues and fix those instead of symptoms?

chavonadelal commented:
WOW

cfif126 commented:
hi plz merge

chavonadelal commented on Content.Shared/Construction/Prototypes/ConstructionPrototype.cs:
```diff
@@ -14,89 +12,80 @@ public sealed partial class ConstructionPrototype : IPrototype
     /// </summary>
     [DataField("description")]
-    public string Description = string.Empty;
+    public LocId? DescLocId;
+
+    public string? Description;
```
```suggestion
    public LocId? Description;
```

Use only the LocID field so that your work doesn't go down the drain when someone decides to add non-localized text.

chavonadelal commented on Content.Shared/Construction/Prototypes/ConstructionPrototype.cs:
```diff
@@ -14,89 +12,80 @@ public sealed partial class ConstructionPrototype : IPrototype
     /// </summary>
     [DataField("name")]
-    public string Name = string.Empty;
+    public LocId? NameLocId;
+
+    public string? Name;
```
```suggestion
    public LocId? Name;
```
Use only the LocID field so that your work doesn't go down the drain when someone decides to add non-localized text.

chavonadelal reviewed (changes requested):
Use only the LocID fields

EmoGarbage404 commented:
there's a small visual issue but it's related to the engine-side UI code. I'll put up a fix shortly.

ScarKy0 commented:
I'll look at it in-game later and merge if its fine
If another maintainer wants to merge before me feel free
